File: integration-kits/ts/src/in_memory_client_host.test.ts
import { describe, it, expect, beforeEach, beforeAll } from "vitest";
import { InMemorySpilmanClientHost } from "./in_memory_client_host.js";
import { init } from "./index.js";
import * as secp from "@noble/secp256k1";

// Initialize WASM before running any tests
beforeAll(async () => {
  await init();
});

// Generate a test keypair
function generateTestKeypair(): { secretHex: string; pubkeyHex: string } {
  const secretBytes = secp.utils.randomPrivateKey();
  const pubkeyBytes = secp.getPublicKey(secretBytes, true);
  return {
    secretHex: Buffer.from(secretBytes).toString("hex"),
    pubkeyHex: Buffer.from(pubkeyBytes).toString("hex"),
  };
}

describe("InMemorySpilmanClientHost", () => {
  let host: InMemorySpilmanClientHost;
  let keypair: { secretHex: string; pubkeyHex: string };

  beforeEach(() => {
    keypair = generateTestKeypair();
    host = new InMemorySpilmanClientHost(keypair.secretHex);
  });

  describe("funding storage", () => {
    it("getChannelFunding returns null for unknown channel", () => {
      expect(host.getChannelFunding("unknown")).toBeNull();
    });

    it("saveOpeningChannel stores data and getChannelFunding retrieves it", () => {
      const channelId = "test-channel-123";
      const fundingJson = '{"params_json":"{}","funding_proofs_json":"[]"}';

      host.saveOpeningChannel(channelId, fundingJson);
      expect(host.getChannelFunding(channelId)).toBe(fundingJson);
    });

    it("saveOpeningChannel sets state to opening", () => {
      const channelId = "test-channel-456";
      host.saveOpeningChannel(channelId, "{}");
      expect(host.getChannelState(channelId)).toBe("opening");
    });

    it("markChannelOpen transitions state to open", () => {
      const channelId = "test-channel-789";
      host.saveOpeningChannel(channelId, "{}");
      expect(host.getChannelState(channelId)).toBe("opening");

      host.markChannelOpen(channelId, "[]");
      expect(host.getChannelState(channelId)).toBe("open");
    });
  });

  describe("payment state", () => {
    it("getPaymentState returns null initially", () => {
      expect(host.getPaymentState("any-channel")).toBeNull();
    });

    it("recordPayment stores state and getPaymentState retrieves it", () => {
      const channelId = "test-channel";
      const stateJson = '{"balance":100,"signature":"abc123"}';

      host.recordPayment(channelId, stateJson);
      expect(host.getPaymentState(channelId)).toBe(stateJson);
    });

    it("recordPayment updates existing state", () => {
      const channelId = "test-channel";
      host.recordPayment(channelId, '{"balance":100}');
      host.recordPayment(channelId, '{"balance":200}');
      expect(host.getPaymentState(channelId)).toBe('{"balance":200}');
    });
  });

  describe("lifecycle", () => {
    it("getChannelState returns 'open' for unknown channel", () => {
      expect(host.getChannelState("unknown")).toBe("open");
    });

    it("markChannelClosed changes state to 'closed'", () => {
      const channelId = "test-channel";
      host.saveOpeningChannel(channelId, "{}");
      host.markChannelOpen(channelId, "[]");
      expect(host.getChannelState(channelId)).toBe("open");

      host.markChannelClosed(channelId);
      expect(host.getChannelState(channelId)).toBe("closed");
    });

    it("listChannelIds returns all stored channels", () => {
      host.saveOpeningChannel("channel-1", "{}");
      host.saveOpeningChannel("channel-2", "{}");

      const ids = host.listChannelIds();
      expect(ids).toHaveLength(2);
      expect(ids).toContain("channel-1");
      expect(ids).toContain("channel-2");
    });

    it("deleteChannel removes channel and all its data", () => {
      const channelId = "test-channel";
      host.saveOpeningChannel(channelId, '{"funding":true}');
      host.recordPayment(channelId, '{"payment":true}');
      host.markChannelClosed(channelId);

      // Verify data exists
      expect(host.getChannelFunding(channelId)).not.toBeNull();
      expect(host.getPaymentState(channelId)).not.toBeNull();
      expect(host.getChannelState(channelId)).toBe("closed");
      expect(host.listChannelIds()).toContain(channelId);

      // Delete
      host.deleteChannel(channelId);

      // Verify data is gone
      expect(host.getChannelFunding(channelId)).toBeNull();
      expect(host.getPaymentState(channelId)).toBeNull();
      expect(host.getChannelState(channelId)).toBe("open"); // Default for unknown
      expect(host.listChannelIds()).not.toContain(channelId);
    });
  });

  describe("time", () => {
    it("nowSeconds returns reasonable Unix timestamp", () => {
      const before = BigInt(Math.floor(Date.now() / 1000));
      const got = host.nowSeconds();
      const after = BigInt(Math.floor(Date.now() / 1000));

      expect(got).toBeGreaterThanOrEqual(before);
      expect(got).toBeLessThanOrEqual(after);
    });
  });

  describe("crypto", () => {
    it("computeChannelSecret produces deterministic result", () => {
      const receiver = generateTestKeypair();

      const secret1 = host.computeChannelSecret(keypair.pubkeyHex, receiver.pubkeyHex);
      const secret2 = host.computeChannelSecret(keypair.pubkeyHex, receiver.pubkeyHex);

      expect(secret1).toBe(secret2);
      expect(secret1).toHaveLength(64); // 32 bytes hex encoded
    });

    it("signWithTweakedKey produces valid signature", () => {
      const messageHex = "0000000000000000000000000000000000000000000000000000000000000001";
      const tweakHex = "0000000000000000000000000000000000000000000000000000000000000002";

      const sig = host.signWithTweakedKey(keypair.pubkeyHex, messageHex, tweakHex);
      expect(sig).toHaveLength(128); // 64 bytes hex encoded
    });

    it("signWithTweakedKey produces valid hex", () => {
      const messageHex = "0000000000000000000000000000000000000000000000000000000000000001";
      const tweakHex = "0000000000000000000000000000000000000000000000000000000000000002";

      const sig = host.signWithTweakedKey(keypair.pubkeyHex, messageHex, tweakHex);

      // Verify it's valid lowercase hex
      expect(sig).toMatch(/^[0-9a-f]{128}$/);
    });
  });
});

File: crates/cdk-spilman-go/spilman/in_memory_client_host_test.go
//go:build spilman_dev

package spilman

import (
	"testing"
	"time"
)

func TestInMemoryClientHost_FundingStorage(t *testing.T) {
	secret, _, err := GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair failed: %v", err)
	}
	host := NewInMemoryClientHost(secret)

	channelID := "test-channel-123"
	fundingJSON := `{"params_json":"{}","funding_proofs_json":"[]"}`

	// GetChannelFunding returns empty string for unknown channel
	if got := host.GetChannelFunding("unknown"); got != "" {
		t.Errorf("GetChannelFunding(unknown) = %q, want empty string", got)
	}

	// SaveOpeningChannel stores data
	host.SaveOpeningChannel(channelID, fundingJSON)

	// GetChannelFunding retrieves it
	if got := host.GetChannelFunding(channelID); got != fundingJSON {
		t.Errorf("GetChannelFunding() = %q, want %q", got, fundingJSON)
	}

	// SaveOpeningChannel sets state to "opening"
	if got := host.GetChannelState(channelID); got != "opening" {
		t.Errorf("GetChannelState() after save = %q, want %q", got, "opening")
	}

	// MarkChannelOpen transitions state to "open"
	host.MarkChannelOpen(channelID, "[]")
	if got := host.GetChannelState(channelID); got != "open" {
		t.Errorf("GetChannelState() after MarkChannelOpen = %q, want %q", got, "open")
	}
}

func TestInMemoryClientHost_PaymentState(t *testing.T) {
	secret, _, err := GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair failed: %v", err)
	}
	host := NewInMemoryClientHost(secret)

	channelID := "test-channel-456"
	stateJSON := `{"balance":100,"signature":"abc123"}`

	// GetPaymentState returns empty string initially
	if got := host.GetPaymentState(channelID); got != "" {
		t.Errorf("GetPaymentState() initially = %q, want empty string", got)
	}

	// RecordPayment stores state
	host.RecordPayment(channelID, stateJSON)

	// GetPaymentState retrieves it
	if got := host.GetPaymentState(channelID); got != stateJSON {
		t.Errorf("GetPaymentState() = %q, want %q", got, stateJSON)
	}

	// Update with new state
	newStateJSON := `{"balance":200,"signature":"def456"}`
	host.RecordPayment(channelID, newStateJSON)

	if got := host.GetPaymentState(channelID); got != newStateJSON {
		t.Errorf("GetPaymentState() after update = %q, want %q", got, newStateJSON)
	}
}

func TestInMemoryClientHost_Lifecycle(t *testing.T) {
	secret, _, err := GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair failed: %v", err)
	}
	host := NewInMemoryClientHost(secret)

	channelID1 := "channel-1"
	channelID2 := "channel-2"

	// GetChannelState returns "open" for unknown channel
	if got := host.GetChannelState("unknown"); got != "open" {
		t.Errorf("GetChannelState(unknown) = %q, want %q", got, "open")
	}

	// Save two channels
	host.SaveOpeningChannel(channelID1, `{"id":1}`)
	host.MarkChannelOpen(channelID1, "[]")
	host.SaveOpeningChannel(channelID2, `{"id":2}`)
	host.MarkChannelOpen(channelID2, "[]")
	host.RecordPayment(channelID1, `{"balance":10}`)

	// ListChannelIDs returns all stored channels
	ids := host.ListChannelIDs()
	if len(ids) != 2 {
		t.Errorf("ListChannelIDs() returned %d items, want 2", len(ids))
	}

	// Both channels should be in the list (order not guaranteed)
	found1, found2 := false, false
	for _, id := range ids {
		if id == channelID1 {
			found1 = true
		}
		if id == channelID2 {
			found2 = true
		}
	}
	if !found1 || !found2 {
		t.Errorf("ListChannelIDs() = %v, want both %q and %q", ids, channelID1, channelID2)
	}

	// MarkChannelClosed changes state to "closed"
	host.MarkChannelClosed(channelID1)
	if got := host.GetChannelState(channelID1); got != "closed" {
		t.Errorf("GetChannelState() after close = %q, want %q", got, "closed")
	}

	// Other channel still open
	if got := host.GetChannelState(channelID2); got != "open" {
		t.Errorf("GetChannelState(channel2) = %q, want %q", got, "open")
	}

	// DeleteChannel removes channel and all its data
	host.DeleteChannel(channelID1)

	if got := host.GetChannelFunding(channelID1); got != "" {
		t.Errorf("GetChannelFunding() after delete = %q, want empty", got)
	}
	if got := host.GetPaymentState(channelID1); got != "" {
		t.Errorf("GetPaymentState() after delete = %q, want empty", got)
	}

	// ListChannelIDs should only return channel2 now
	ids = host.ListChannelIDs()
	if len(ids) != 1 || ids[0] != channelID2 {
		t.Errorf("ListChannelIDs() after delete = %v, want [%q]", ids, channelID2)
	}
}

func TestInMemoryClientHost_NowSeconds(t *testing.T) {
	secret, _, err := GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair failed: %v", err)
	}
	host := NewInMemoryClientHost(secret)

	before := uint64(time.Now().Unix())
	got := host.NowSeconds()
	after := uint64(time.Now().Unix())

	if got < before || got > after {
		t.Errorf("NowSeconds() = %d, want between %d and %d", got, before, after)
	}
}

func TestInMemoryClientHost_Crypto(t *testing.T) {
	secret, pubkey, err := GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair failed: %v", err)
	}
	host := NewInMemoryClientHost(secret)

	// Generate another keypair for the receiver
	_, receiverPubkey, err := GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair (receiver) failed: %v", err)
	}

	// ComputeChannelSecret produces deterministic result
	secret1, err := host.ComputeChannelSecret(pubkey, receiverPubkey)
	if err != nil {
		t.Fatalf("ComputeChannelSecret failed: %v", err)
	}
	if len(secret1) != 64 { // 32 bytes hex encoded
		t.Errorf("ComputeChannelSecret returned %d chars, want 64", len(secret1))
	}

	// Same inputs produce same output
	secret2, err := host.ComputeChannelSecret(pubkey, receiverPubkey)
	if err != nil {
		t.Fatalf("ComputeChannelSecret (2nd call) failed: %v", err)
	}
	if secret1 != secret2 {
		t.Errorf("ComputeChannelSecret not deterministic: %q != %q", secret1, secret2)
	}

	// SignWithTweakedKey produces a signature
	messageHex := "0000000000000000000000000000000000000000000000000000000000000001"
	tweakHex := "0000000000000000000000000000000000000000000000000000000000000002"

	sig, err := host.SignWithTweakedKey(pubkey, messageHex, tweakHex)
	if err != nil {
		t.Fatalf("SignWithTweakedKey failed: %v", err)
	}
	if len(sig) != 128 { // 64 bytes hex encoded
		t.Errorf("SignWithTweakedKey returned %d chars, want 128", len(sig))
	}

	// Note: Schnorr signatures may use random nonces, so we just verify
	// that it produces a valid-looking signature (correct length, valid hex)
	for _, c := range sig {
		if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')) {
			t.Errorf("SignWithTweakedKey returned invalid hex character: %c", c)
		}
	}
}
